Engineering/Mobile/ios_bluetooth_low_energy/ios_bluetooth_low_energy.md

```diff
@@ -0,0 +1,376 @@
+
+* Central scans for Peripheral and connects to it
+* Central reads a value from Peripheral
+* Central writes a value to Peripheral
+* Peripheral notifies Central that a value has changed
+
+Each BLE Central is compatible with each BLE Peripheral, because they use the same service and characteristics UUIDs.
+
+### Development tools
+
+* Xcode - for iOS project
+
+## Supported devices
+
+Platform | Peripheral | Central | Notes
+----- | ----- | ----- | -----
+iOS | YES | YES | iOS 11.0 and newer
+
+## Table of UUIDs
+
+Name | UUID
+----- | ---------------
+Service | 000018aa-0000-1000-8000-00805f9b34fb
+Characteristic for read | 00002bb0-0000-1000-8000-00805f9b34fb
+Characteristic for write | 00002bb2-0000-1000-8000-00805f9b34fb
+Characteristic for indicate | 00002bb1-0000-1000-8000-00805f9b34fb
+
+## BLE Central (all platforms)
+
+Central (also called Master or Client) works similarly on all platforms:
+
+* Scans for Peripherals which have a service with our UUID (see Table of UUIDs)
+* Connects to the first found Peripheral
+* Discovers services and characteristics of the connected Peripheral
+* Subscribes to indications of the "Characteristic for indicate" (see Table of UUIDs)
+* Allows the user to disconnect from Peripheral
+* Allows the user to read a characteristic value (string) from Peripheral
+* Allows the user to write a characteristic value (string) to Peripheral
+
+The App will start the central, discover the services + Characteristics, connect to the peripheral, read a value, request to be notified for value update, write a value.
+
+I have attached the DEMO source code of 2 project. The Central source simulate the iOS Device and the Peripheral source simulate the Bluetooth device with specify service and characteristic in the Table UUID.
+
+We have a lot of functions detail in the source code and here are some important functions I want to note with you.
+
+## IN BLEProofCentral SOURCE
+
+* App request permission for Bluetooth event. for that, in info.plist, I added this :
+
+```
+    <key>NSBluetoothAlwaysUsageDescription</key>
+    <string>We use Bluetooth to show basic communication between Central and Peripheral</string>
+    <key>NSBluetoothPeripheralUsageDescription</key>
+    <string>We use Bluetooth to show basic communication between Central and Peripheral</string>
+    <key>UIApplicationSupportsIndirectInputEvents</key>
+```
+
+* Init BLE
+
+```
+    private func initBLE() {
+        // using DispatchQueue.main means we can update UI directly from delegate methods
+        bleCentral = CBCentralManager(delegate: self, queue: DispatchQueue.main)
+    }
+```
+
+* BLELifecycleState
+
+```
+    enum BLELifecycleState: String {
+    case bluetoothNotReady
+    case disconnected
+    case scanning
+    case connecting
+    case connectedDiscovering
+    case connected
+    }
+```
+
+When scan and connect the central to peripheral, implement count down 30s timeout.
+
+* Count down time
+
+```
```

Comment:
Is this swift language? If so, can you specify `lang` of the block for the sake of sytax highlight?

```suggestion
```swift
```